### PR TITLE
#161966914 Implement social login functionalities

### PR DESCRIPTION
**Pull request by PromasterGuru**
#### What does this PR do?
- Implement social login with Google, Facebook, and Twitter

#### Description of Task to be completed?
- Implement social login functionalities for oauth1
- Implement social login functionalities for oauth2
- Refactor social login tests to pass

#### How should this be manually tested?
   `$ https://github.com/andela/ah-infinity-stones.git`
   `$ cd ah-infinity-stones`
  ` $ git checkout ft-social-login-161966914`
   `$ virutalenv -m python3 venv`
   `$ source venv/bin/activate`
   `$ pip install -r requirements.txt`

Create a .env file and add the relevant environmental variables
```
export DATABASE_NAME='db_name'
export DATABASE_TEST='test_db_name'
export DATABASE_USER='db_user_name'
export DATABASE_PASSWORD='db_user_password'
export DATABASE_HOST=''
export DATABASE_PORT='5432'
export SECRET_KEY='your_secret_key'
export GOOGLE_KEY='google_key'
export GOOGLE_SECRET_KEY='google_secret_key'
export FACEBOOK_KEY='facebook_app_id'
export FACEBOOK_SECRET_KEY='facebook_user_secret_key'
export TWITTER_KEY='twitter_key'
export TWITTER_SECRET_KEY='twitter_secret_key'
export EMAIL_HOST_USER='email_host_user'
export EMAIL_HOST_PASSWORD='_email_host_user_password'
```

Start postman and use the following url 
`http://127.0.0.1:8000/api/login/oauth/`

Use the following links to get your access tokens
[Google](https://developers.google.com/oauthplayground/):
```
{
    "provider":"google-oauth2",
    "access_token":"valid_google_access_token"
}
```

[Facebook](https://developers.facebook.com/tools/explorer/)
```
{
    "provider":"facebook",
    "access_token":"valid_facebook_access_token"
}
```

[Twitter](https://developer.twitter.com/en/dashboard)
```
{
    "provider": "twitter",
    "access_token": "valid_access_token",
    "access_token_secret": "valid_access_token_secret"

}
```

#### What are the relevant pivotal tracker stories?
[#161966914](https://www.pivotaltracker.com/story/show/161966914)

#### Screenshots (if appropriate)

screenshot 2018-12-17 at 20 30 13
<br /><br />
screenshot 2018-12-17 at 22 17 04
<br /><br />
screenshot 2018-12-17 at 22 18 55
<br /><br />
screenshot 2018-12-17 at 20 34 46

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,6 +4,7 @@ __pycache__/
 *$py.class
 .vscode/
 .DS_Store
+/.DS_Store
```
Comment (r-wambui) [inline .gitignore]: Why do we have 2 .DS_Store files?

File: requirements.txt [hunk 2]
```diff
@@ -27,6 +27,7 @@ pycodestyle==2.4.0
 pyflakes==2.0.0
 PyJWT==1.7.0
 pylint==2.2.2
+python-social-auth==0.3.6
```
Comment (r-wambui) [inline requirements.txt]: Do we require this? or are we using `social-auth-app-django==3.1.0`?

File: authors/apps/authentication/tests/test_registration.py [hunk 3]
```diff
@@ -43,7 +43,8 @@ def test_send_verification_mail(self):
         self.assertEqual(mail.outbox[0].subject, "Activate your account.")
         self.assertEqual(response.status_code, status.HTTP_201_CREATED)
 
-    def test_empty_username_field(self):
+
```
Comment (r-wambui) [inline authors/apps/authentication/tests/test_registration.py]: The test is failing because of some changes you made here. Please check.

Comment (r-wambui) [reply authors/apps/authentication/tests/test_registration.py]: Check on this `IndentationError: unindent does not match any outer indentation level`

File: .gitignore (hunk 1 above)
Comment (PromasterGuru) [reply .gitignore]: The first .DS_Store file ignores .DS_Store in the root folder while the second one ignores .DS_Store files inside subfolders.

Comment (PromasterGuru) [reply .gitignore]: I think the second one wasn't neccessary, I have eliminated it.

File: requirements.txt (hunk 2 above)
Comment (PromasterGuru) [reply requirements.txt]: Actually, it is not necessary. I have removed it.

File: .gitignore (hunk 1 above)
Comment (amutava) [reply .gitignore]: @PromasterGuru check how to ignore from all the folders and add that to this file.

**Comment by r-wambui** [comment]
@PromasterGuru Please check that the travis completes the build.

**Pull request merged**